### PR TITLE
Implement set time zone and display time in the correct time zone

### PR DESCRIPTION
**Pull request by allenwq**
For #1133 

User now is able to set their time zone in account settings. 

Time will be displayed in user's current time zone, it will fall back to the default one in the configuration if not set.

BTW, I haven't investigated why the format of schema.rb changed after `db:migrate`, I suspect it's due to some gem update. Is it same for all of you guys ? @kxmbrian @weiqingtoh

### CODE REVIEW

**Comment by allenwq** [comment]
Don't review yet, will check the schema file.

**Comment by allenwq** [comment]
OK I did the migration again, schema file's format still changed.

**Comment by weiqingtoh** [comment]
LGTM. @kxmbrian ?

File: app/controllers/concerns/application_user_time_zone_concern.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+  included do
+    around_action :set_time_zone, if: :current_user
+  end
+
+  protected
+
+  # Set the time_zone for current request.
+  def set_time_zone(&block) # rubocop:disable Style/AccessorMethodName
+    Time.use_zone(current_user.time_zone || Application.config.x.user_default_time_zone, &block)
```
Comment (lowjoel) [inline app/controllers/concerns/application_user_time_zone_concern.rb]: default_user_time_zone sounds more natural.

**Comment by lowjoel** [comment]
@allenwq https://github.com/SchemaPlus/schema_plus/issues/228

File: app/controllers/concerns/application_user_time_zone_concern.rb (hunk 1 above)
Comment (allenwq) [inline app/controllers/concerns/application_user_time_zone_concern.rb]: Renamed, Thanks !

File: app/controllers/concerns/application_user_time_zone_concern.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+module ApplicationUserConcern
```
Comment (kxmbrian) [inline app/controllers/concerns/application_user_time_zone_concern.rb]: Is it better to exclude this file from this commit?

Comment (allenwq) [inline app/controllers/concerns/application_user_time_zone_concern.rb]: Thanks ! Didn't notice this one got checked in and removed ><

**Comment by allenwq** [comment]
git bisect tells me that commit c3bc2b3fa02023215378ef85d256078e866d6ff0 changes the format.

@lowjoel The issue you posted seems to be different ?

File: app/models/user.rb [hunk 3]
```diff
@@ -38,6 +38,7 @@ def deleted
 
   validates :email, :encrypted_password, :authentication_token, absence: true, if: :built_in?
   schema_validations except: [:encrypted_password]
+  validates :time_zone, inclusion: { in: ActiveSupport::TimeZone.zones_map.keys }, allow_nil: true
```
Comment (lowjoel) [inline app/models/user.rb]: what happens if time zones are removed from this list?

Comment (allenwq) [inline app/models/user.rb]: In that case user will see a validation error when update the profile.

Comment (lowjoel) [inline app/models/user.rb]: I'm aware. What about in the time between the timezone is removed and the user not updating his profile? What time zone will he get?

Comment (allenwq) [inline app/models/user.rb]: Yes the exception will be raised, only that user will be affected. Yah I didn't handle that case, thanks I will add an issue regarding this.

**Pull request merged**